onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc

```diff
@@ -831,23 +831,74 @@ onnxruntime::Node& NodeFromApiNode(onnx_layout_transformation::api::NodeRef& nod
   return static_cast<ApiNode&>(node).Node();
 }
 
+CostCheckResult OrtDefaultCostCheck(const api::GraphRef& graph, const api::NodeRef& node,
+                                    const std::vector<int64_t>& /*perm*/,
+                                    const std::unordered_set<std::string>& /*outputs_leading_to_transpose*/) {
+  // special case some kernels based on the ORT implementation details
+  if (node.GetExecutionProviderType() == kCpuExecutionProvider) {
```

Comment:
Shall we have similar clues for CUDA EP? For example, In Volta or later GPU (SM >= 70), NHWC convolution might be faster than NCHW convolution (cudNN internally convert NCHW to NHWC in many cases).

Reply:
We can certainly add them here. I don't know what the rules should be, but we now have an ORT specific place for these sorts of things to be incorporated.